### PR TITLE
[LLVM] Remove nuw neg

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch removes APIs that creating NUW neg. It is a trivial case because `sub nuw 0, X` always gets simplified into zero.
I believe there is no optimization opportunities in the real-world applications that we can take advantage of the nuw flag.

Motivated by https://github.com/llvm/llvm-project/pull/84792#discussion_r1524891134.

Compile-time improvement: https://llvm-compile-time-tracker.com/compare.php?from=d1f182c895728d89c5c3d198b133e212a5d9d4a3&to=da7b7478b7cbb32c09d760f6b8d0e67901e0d533&stat=instructions:u

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 1]
```diff
@@ -4250,10 +4250,11 @@ static Instruction *canonicalizeAbs(BinaryOperator &Xor,
     // xor (add A, Op1), Op1  ; add -1 and flip bits if negative
     // --> (A < 0) ? -A : A
     Value *IsNeg = Builder.CreateIsNeg(A);
-    // Copy the nuw/nsw flags from the add to the negate.
+    // Copy the nsw flags from the add to the negate.
     auto *Add = cast<BinaryOperator>(Op0);
-    Value *NegA = Builder.CreateNeg(A, "", Add->hasNoUnsignedWrap(),
-                                   Add->hasNoSignedWrap());
+    Value *NegA = Add->hasNoUnsignedWrap()
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4255]: This change fixes a regression caused by not propagating nuw flags.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4255]: Why is dropping the `nuw` APIs from neg related to this fixing the bug?Should this be independent commits?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4255]: I don't know if it is OK to drop these changes and leave the regression.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4255]: Oh I see what you mean. I misunderstood and  thought you meant this patch was really all about this little fix here.

**Comment by goldsteinn** [comment]
Why is tramp3d-v4 sped up so so much in comparison to the surrounding cases?
I guess the compile time advantage probably justifies this change though.

**Comment by goldsteinn** [comment]
LGTM.

Wait for 1 more signoff.

My only concern is there is some case we prove the `nuw` (or are carrying around a `nuw`) without inspecting the operands, but I guess if there where zero changes on your benchmark thats unfounded.

File: llvm/lib/IR/Core.cpp [hunk 2]
```diff
@@ -3555,11 +3550,6 @@ LLVMValueRef LLVMBuildNSWNeg(LLVMBuilderRef B, LLVMValueRef V,
   return wrap(unwrap(B)->CreateNSWNeg(unwrap(V), Name));
 }
 
-LLVMValueRef LLVMBuildNUWNeg(LLVMBuilderRef B, LLVMValueRef V,
-                             const char *Name) {
-  return wrap(unwrap(B)->CreateNUWNeg(unwrap(V), Name));
-}
```
Comment (nikic) [inline llvm/lib/IR/Core.cpp:3561]: We should avoid dropping C APIs if it's still possible to support them (e.g. by explicitly setting the flag, or even just creating the instruction without the flag).

Comment (dtcxzyw) [reply llvm/lib/IR/Core.cpp:3561]: I will make them deprecated.

Comment (dtcxzyw) [reply llvm/lib/IR/Core.cpp:3561]: Done.

File: llvm/include/llvm-c/Core.h [hunk 3]
```diff
@@ -2316,6 +2316,10 @@ LLVMValueRef LLVMAlignOf(LLVMTypeRef Ty);
 LLVMValueRef LLVMSizeOf(LLVMTypeRef Ty);
 LLVMValueRef LLVMConstNeg(LLVMValueRef ConstantVal);
 LLVMValueRef LLVMConstNSWNeg(LLVMValueRef ConstantVal);
+/**
+ *
+ * @deprecated Use LLVMConstNull instead.
+ */
```
Comment (nikic) [inline llvm/include/llvm-c/Core.h:2322]: There is a deprecation macro for the C API you can use as well.

**Review by nikic — APPROVED** [review]

**Comment by tuliom** [comment]
FYI: this commit caused a regression. Reported as #86944 .

**Pull request merged**